source/develop/release-management/features/sla/hosted-engine-agent-offloading.html.md

```diff
@@ -0,0 +1,53 @@
+
+## Summary
+
+Initially, when hosted engine tools was designed, they were intentionally split to the agent and the broker. Agent planned to be a lightweight application, responsible for cluster management decisions. And broker was supposed to make all the hard work - storage operations, monitoring, locking etc. 
+However, during hosted engine development, agent was burdened with additional load, like OVF extraction, storage preparation and other activities. 
+
+The goal of that feature is to make agent lightweight again and keep only management code in it, while moving all other stuff to the broker. This will result in a more maintenable code and more reliable behaviour of the hosted engine.
+## Owner
+
+*   Name: [Denis Chaplygin](https://github.com/akashihi)
+*   Email: <dchaplyg@redhat.com>
+
+## Detailed Description
+
+The specific goals associated with this feature are as follows:
+
+*   Move storage mounting/unmounting from agent to the broker.
+*   Broker should start/stop domain monitor on agent's command. Domain monitor could be implemented as submonitor.
+    
+    Domain monitoring needs to be changed drastically. When storage validation approach was changed (See [Documentation/External references](develop/release-management/features/sla/hosted-engine-agent-offloading/#documentation)), domain monitoring become an important for domain to be valied. Non-monitored domains are not reported as valid anymore. 
+    Therefore, old approach, when domains monitoring was active only outside of the LocalMaintenance mode is not applicable anymore. I propose it to be changed in the following way - domain monitoring still should be disabled in the LocalMaintenance mode. While domain monitoring is disabled, actual domain validation will be skipped and domain will be reported as valid. Broker will still try to update status (and read statues of the other cluster participants) even in the maintenance mode, so it will have a chance to block on broken storage. Thus, all I/O should be moved to a separate thread, so broker will not block on I/O issues, only that thread.
+*   Move sanlock acquire/release activities form agent to the broker.
+*   As steps above introduce new possible 'half-configured' state, FSM should be extended with that state.
+*   Agent should post state to the broker only and broker must publish it on storage if possible.
+
+    Broker will keep current status in memory and reply to agent, hosted-engine setup and vdsm using that in-memory state. Agent will call broker's RPC to update that state. Upon in-memory state update separate disk writing thread will be notified. That separate disk writing thread will be started during broker start-up sequence and it's only purpose is to write the state to the storage and block in case of storage failures, while letting other broker code to run. I plan to use deque with size one to pass the current state to disk writing thread and use Conditions to synchronize threads.
```

Comment:
- "Broker will keep current status in memory and reply to agent, hosted-engine setup and vdsm using that in-memory state." -> "The broker will keep the current status in memory and reply to the agent, hosted-engine-setup, and VDSM using that in-memory state."
- "Agent will call broker's" -> "The agent will call the broker's"
- "Upon in-memory state update separate disk writing thread will be notified." -> "Upon an in-memory state update, a separate disk-writing thread will be notified."
- "disk writing thread will be started during broker start-up sequence and it's" -> "disk-writing thread will be started during the broker start-up sequence and its"
- "other broker code to run" -> "other broker code run"
- "size one to pass the current state to disk writing thread" -> "a size of one to pass the current state to the disk-writing thread"

Reply:
Done